File: SwipeToAction/src/main/java/com/github/chantsune/swipetoaction/animations/SwipeAnimation.java
package com.github.chantsune.swipetoaction.animations;

import android.view.View;
import android.view.animation.Animation;
import android.view.animation.DecelerateInterpolator;
import android.view.animation.Transformation;

import androidx.core.view.ViewCompat;

import static com.github.chantsune.swipetoaction.extensions.Utils.setViewWidth;

/**
 * Created by Alexander Pataridze
 */

public class SwipeAnimation extends Animation {
    private int width;
    private int startWidth = -1;
    private View resizeView;
    private View changeXView;
    private boolean left;

    public SwipeAnimation(View resizeView, int width, View changeXView, boolean left) {
        this.resizeView = resizeView;
        this.width = width;
        this.changeXView = changeXView;
        this.left = left;
        setDuration(300);
        setInterpolator(new DecelerateInterpolator());
    }


    @Override
    protected void applyTransformation(float interpolatedTime, Transformation t) {

        if (startWidth < 0) {
            startWidth = resizeView.getWidth();
        }

        setViewWidth(resizeView, startWidth + (int) (((float) width - (float) startWidth) * interpolatedTime));

        if (left) {
            ViewCompat.setTranslationX(changeXView, resizeView.getWidth());
        } else {
            ViewCompat.setTranslationX(changeXView, -resizeView.getWidth());
        }

    }

    @Override
    public boolean willChangeBounds() {
        return true;
    }
}

File: SwipeToAction/src/main/java/com/github/chantsune/swipetoaction/animations/SwipeAnimation.kt
package com.github.chantsune.swipetoaction.animations

import android.view.View
import android.view.animation.Animation
import android.view.animation.DecelerateInterpolator
import android.view.animation.Transformation
import com.github.chantsune.swipetoaction.extensions.viewWidth

class SwipeAnimation(
    private val resizeView: View,
    private val width: Int,
    private val changeXView: View,
    private val left: Boolean
) : Animation() {
    private var startWidth = -1

    init {
        duration = 300
        interpolator = DecelerateInterpolator()
    }

    override fun applyTransformation(interpolatedTime: Float, t: Transformation) {
        if (startWidth < 0) {
            startWidth = resizeView.width
        }
        resizeView.viewWidth = startWidth + ((width - startWidth) * interpolatedTime).toInt()

        if (left) {
            changeXView.translationX = resizeView.width.toFloat()
        } else {
            changeXView.translationX = -resizeView.width.toFloat()
        }
    }

    override fun willChangeBounds(): Boolean = true
}
